assets/js/blocks/attribute-filter/block.js

```diff
@@ -0,0 +1,173 @@
+import { ATTRIBUTES } from '@woocommerce/block-settings';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import CheckboxList from '@woocommerce/base-components/checkbox-list';
+
+/**
+ * Component displaying an attribute filter.
+ */
+const AttributeFilterBlock = ( { attributes } ) => {
+	const { showCounts, attributeId, queryType } = attributes;
+
+	const [ options, setOptions ] = useState( [] );
+	const [ checkedOptions, setCheckedOptions ] = useState( [] );
+	const [ currentAttribute, setCurrentAttribute ] = useState( [] );
+
+	const [ productAttributes, setProductAttributes ] = useQueryStateByKey(
+		'product-grid',
+		'attributes',
+		[]
+	);
+
+	const [ queryState ] = useQueryStateContext( 'product-grid' );
+
+	const { results: allTerms, isLoading: allTermsIsLoading } = useCollection( {
+		namespace: '/wc/blocks',
+		resourceName: 'products/attributes/terms',
+		resourceValues: [ attributeId ],
+	} );
+
+	const {
+		results: filteredCounts,
+		isLoading: filteredCountsIsLoading,
+	} = useCollection( {
+		namespace: '/wc/store',
+		resourceName: 'products/collection-data',
+		query: {
+			...queryState,
+			orderby: undefined,
+			order: undefined,
+			per_page: undefined,
+			page: undefined,
+			calculate_attribute_counts: [ attributeId ],
+		},
+	} );
+
+	const getLabel = useCallback(
+		( name, count ) => {
+			return (
+				<Fragment>
+					{ name }
+					{ showCounts && count && (
+						<span className="wc-block-attribute-filter-list-count">
+							{ count }
+						</span>
+					) }
+				</Fragment>
+			);
+		},
+		[ showCounts ]
+	);
+
+	const getFilteredTerm = useCallback(
+		( id ) => {
+			if ( ! filteredCounts.attribute_counts ) {
+				return {};
+			}
+			return filteredCounts.attribute_counts.find(
+				( { term } ) => term === id
+			);
+		},
+		[ filteredCounts ]
+	);
+
+	/**
+	 * Compare intersection of all terms and filtered counts to get a list of options to display.
+	 */
+	useEffect( () => {
+		if ( filteredCountsIsLoading || allTermsIsLoading ) {
+			return;
+		}
+
+		const newOptions = [];
+
+		allTerms.forEach( ( term ) => {
+			const filteredTerm = getFilteredTerm( term.id );
+
+			if ( ! filteredTerm && null !== filteredCounts ) {
+				return;
+			}
+
+			newOptions.push( {
+				key: term.slug,
+				label: getLabel(
+					term.name,
+					filteredTerm ? filteredTerm.count : term.count
+				),
+				checked: false,
+			} );
+		} );
+
+		setOptions( newOptions );
+	}, [
+		filteredCounts,
+		allTerms,
+		showCounts,
+		filteredCountsIsLoading,
+		allTermsIsLoading,
+	] );
+
+	useEffect( () => {
+		setCurrentAttribute(
+			find( ATTRIBUTES, [ 'attribute_id', attributeId + '' ] )
+		);
+	}, [ attributeId ] );
+
+	useEffect( () => {
+		const taxonomy = 'pa_' + currentAttribute.attribute_name;
+
+		if ( ! taxonomy ) {
+			return;
+		}
+
+		const newProductAttributes = productAttributes.filter(
+			( item ) => item.attribute !== taxonomy
+		);
+		const slug = join( checkedOptions, ',' );
```

Comment:
Could we use the native method here?
```suggestion
		const slug = checkedOptions.join( ',' );
```